sshd-common/src/main/java/org/apache/sshd/common/future/DefaultSshFuture.java

```diff
@@ -79,6 +82,9 @@ protected Object await0(long timeoutMillis, boolean interruptable) throws Interr
                 }
 
                 curTime = System.currentTimeMillis();
+                if (curTime >= endTime) {
+                    setTimeout();
+                }
```

Comment:
I think we also need to call setTimeout() above, in the exception handler, inside `if (interruptable) {`.